wgengine/magicsock/magicsock.go

```diff
@@ -1887,6 +1887,11 @@ func (c *Conn) Close() error {
 		c.pconn6.Close()
 	}
 	err := c.pconn4.Close()
+	// activeDerp is non-nil iff at least one connection was opened.
+	// We must wait on derpStarted to ensure the goroutine running dc.Connect has exited.
```

Comment:
comment still doesn't say why.

You can file a public bug with the data race and reference the issue number in the comment.

I'm guessing what the real problem/race might be, but if it's Connect calling back into Conn, perhaps that method that it calls back into needs to check `c.closed`?

Reply:
This _seems_ to be a case of `derphttp.Client` not checking context cancellation frequently enough (e.g. it uses `http.NewRequest` vs `http.NewRequestWithContext`). Let me try improving that and seeing if it gets us back within the resource check timeout (100ms). This is supported by the fact that bumping the timeout to 200ms also fixes the issue.

Reply:
Looking more carefully, no, it seems to be plumbed around extensively, and `NewRequestWithContext` is useless there. In any case, issue created.

Reply:
Ah, this is a goroutine leak test failure, not a data race.

The comment could be like:

  // Our tests verify we don't leak goroutines, so make sure all the DERP
  // goroutines have shut down before Close returns:

Reply:
The derphttp.Client connect path is full of magic. I think it all looks fine and it does plumb contexts around properly everywhere.

I think you fix here is fine, except for the comment.